### DISCUSSION TITLE
Improve detection of nuget source URL from Artifactory

### DISCUSSION THREAD

**Discussion by fgreinacher** [renovatebot/renovate#28041] [Suggest an Idea]
### Tell us more.

Renovate currently tries to derive the source URL of a NuGet dependency via the [`PackageBaseAddress` resource](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource), see https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/nuget/v3.ts#L203-L214. 

This works well for packages from NuGet.org, but fails for Artifactory (and maybe others) because it does not provide this resource.

This issue has just recently been [solved by Dependabot](https://github.com/dependabot/dependabot-core/pull/9117) and I suggest we extend the current logic in a similar way.

I have set up a reproduction at https://github.com/fgreinacher/renovate-artifactory-source-url:

- PR using NuGet.org correctly shows the source URL and the changelog: https://github.com/fgreinacher/renovate-artifactory-source-url/pull/3
- PR using Artifactory is missing both: https://github.com/fgreinacher/renovate-artifactory-source-url/pull/2

**Comment by rarkins** [comment]
Reproduction forked to https://github.com/renovate-reproductions/28041 (and renovate.json syntax error fixed)

**Reply by fgreinacher** [reply]
Thanks 😀 Just be aware that this uses a 30 day trial Artifactory that I have just set up.

**Comment by rarkins** [comment]
Dependabot is not open source so we should not use their _code_ in any way, ideally solve it independently/from first principles.

@fgreinacher is the github source repo contained in any of the JSON responses from JFrog?

Or is the idea to assume that the sourceUrl for package X on an alternative host is the same sourceUrl as package X on the default registry?

**Reply by fgreinacher** [reply]
> is the github source repo contained in any of the JSON responses from JFrog?

@rarkins The source repo is only available in the `nuspec` file. For NuGet.org we can guess it's URL from the package base address (this is not documented though), for Artifactory (and probably other API implementations) we need to download the package and extract the `nuspec` from there.

> Dependabot is not open source so we should not use their code in any way, ideally solve it independently/from first principles.

Yes sure, I just wanted to highlight that this is a feasible solution.

**Reply by viceice** [reply]
I don't like to download the nupkg files, this would make too much trafic. the nupkg files are usally huge

**Reply by viceice** [reply]
We should try to upstream issues for all custom nuget implementaions, so they implement the proper apis. we recently had a similar discussion and the provider has implemented the needed changes within a week.

**Reply by rarkins** [reply]
You evidently have never experienced JFrog before ;)

**Reply by fgreinacher** [reply]
I agree that upstream support would be great. I'll reach out to them, but I'm also not too confident that this will bring the desired outcome :)

**Reply by rarkins** [reply]
If it is just for package metadata which rarely changes then we can cache it very aggressively too, e.g. only check for changes weekly or monthly. How big is the download for this example package for example?

**Reply by fgreinacher** [reply]
This one is 2.33 MB and I guess that's more than the mean. There might be some outliers for sure, not sure whether/how we can handle these.

I have posted https://stackoverflow.com/questions/78192777/download-nuspec-via-nuget-v3-api and also reached out via our official support channels, let's see how that goes.

Maybe we could make it opt-in via an experimental flag to start with?

**Reply by rarkins** [reply]
To give this value some perspective, https://registry.npmjs.org/react is ~6MB _compressed_ (over the wire), while renovate is ~75MB. So I don't think that this size is alone a reason to avoid this, especially as it will be much more cacheable than these two examples

**Reply by viceice** [reply]
I'm ok with opt-in. Will need this disabled, because it would overload our forgejo / s3 again (We'll have huge packages which will have releases much frequently, so caching will not solve it). Forgejo will support that api with next major release 💪 .

AS workaround I'm currently simply using a preset which overrides `sourceUrl` via package rules.

- https://github.com/visualon/renovate-config/blob/main/meta.json#L7

**Reply by rarkins** [reply]
We don't do etag type caching, which would require downloading it anytime the nuspec changed, but instead "once you've checked the nuspec for a sourceUrl then don't check again for a month"

**Reply by fgreinacher** [reply]
I'll send a PR so that we discuss this.

Do you want to make a feature issue out-of-this discussion @rarkins @viceice?

**Reply by rarkins** [reply]
Yes I was planning to, thanks

**Reply by viceice** [reply]
> We don't do etag type caching, which would require downloading it anytime the nuspec changed, but instead "once you've checked the nuspec for a sourceUrl then don't check again for a month"

that file is version specific, so on every new release we need to fetch again.

**Reply by rarkins** [reply]
I would still choose to cache the _result_ for a long period, because it's unlikely to change

**Comment by fgreinacher** [comment]
This is the upstream Artifactory ticket: https://jfrog.atlassian.net/browse/RTFACT-26507

We gave it a little push from our Artifactory team, let's see how it goes.
